src/lang/eval/StateIPCClient.ml

```diff
@@ -0,0 +1,194 @@
+  Unix.connect socket ~addr:(Unix.ADDR_UNIX socket_addr);
+  let (ic, oc) = Unix.in_channel_of_descr socket, Unix.out_channel_of_descr socket in
+  let msg_buf = Jsonrpc.string_of_call ~version: Jsonrpc.V2 call in
+  DebugMessage.plog (Printf.sprintf "Sending: %s\n" msg_buf);
+  (* Send data to the socket. *)
+  let _ = send_delimited oc msg_buf in
+  (* Get response. *)
+  let response = Caml.input_line ic in
+  Unix.close socket;
+  DebugMessage.plog (Printf.sprintf "Response: %s\n" response);
+  Jsonrpc.response_of_string response
+
+(* Encode a literal into bytes, opaque to the backend storage. *)
+let serialize_literal l = Bytes.of_string (PrettyPrinters.literal_to_jstring l)
+let deserialize_literal s tp =
+  try
+    pure @@ ContractState.jstring_to_literal s tp
+  with
+  | Invalid_json s ->
+    fail (s @ mk_error0 "StateIPCClient: Error deserializing literal fetched from IPC call")
+
+(* Map fields are serialized into Ipcmessage_types.MVal
+   Other fields are serialized using serialize_literal into bytes/string. *)
+let rec serialize_field value =
+  match value with
+  | Map (_, mlit) ->
+    let mpb = Caml.Hashtbl.fold (fun key value acc ->
+      let key' = Bytes.to_string (serialize_literal key) in
+      (* values can be Maps or non-map literals. Hence a recursive call. *)
+      let val' = serialize_field value in
+      (key', val') :: acc
+    ) mlit [] in
+    Ipcmessage_types.Mval({ Ipcmessage_types.m = mpb })
+    (* If there are maps _inside_ a non-map field, they are treated same
+     * as non-Map field values are not serialized as protobuf maps. *)
+  | _ -> Ipcmessage_types.Bval (serialize_literal value)
+
+(* Deserialize proto_scilla_val, given its type. *)
+let rec deserialize_value value tp =
+  match value with
+  | Ipcmessage_types.Bval s ->
+    deserialize_literal (Bytes.to_string s) tp
+  | Ipcmessage_types.Mval m ->
+    (match tp with
+    | MapType (kt, vt) ->
+      let mlit = Caml.Hashtbl.create (List.length m.m) in
+      let _ = iterM m.m ~f:(fun (k, v) ->
+        let%bind k' = deserialize_literal k kt in
+        let%bind v' = deserialize_value v vt in
+        Caml.Hashtbl.add mlit k' v';
+        pure ()
+      ) in
+      pure (Map ((kt, vt), mlit))
+    | _ -> fail0 "StateIPCClient: Type mismatch deserializing value. Unexpected protobuf map.")
+
+let encode_serialized_value value =
+  let encoder = Pbrt.Encoder.create () in
+  Ipcmessage_pb.encode_proto_scilla_val value encoder;
+  Bytes.to_string @@ Pbrt.Encoder.to_bytes encoder
+
+let decode_serialized_value value =
+  let decoder = Pbrt.Decoder.of_bytes value in
+  Ipcmessage_pb.decode_proto_scilla_val decoder
+
+let encode_serialized_query query =
+  let encoder = Pbrt.Encoder.create () in
+  Ipcmessage_pb.encode_proto_scilla_query query encoder;
+  Bytes.to_string @@ Pbrt.Encoder.to_bytes encoder
+
+(* Fetch a field value. keys is empty iff this value being fetched is not a whole map itself.
```

Comment:
The second sentence is unrelated to the first one. It (the second sentence of the comment) is relevant only when we're talking about a non-empty `keys` argument to the function.

Reply:
It's still unclear to me. Would it be correct to say that keys is non-empty iff we are fetching (some sub-value of) a map field, including the entire map?

If `field map : Map String String`, and we are fetching using `x <- map`, what is the value of `keys`?

Reply:
My bad. I only looked at the second sentence when addressing your concern. How about I change the first sentence to this?

>Fetch a field value. "keys" is empty when fetching non-map values or an entire Map field.
If a map key is not found, then None is returned, otherwise (Some value) is returned.

Reply:
> I still have one open question, but it relates to something that most likely works. I just don't understand it.

I've now fixed this. Please have a look.
https://github.com/Zilliqa/scilla/pull/644/commits/af0053710ad277ba735f5b0b28c3a5928e606d43

Reply:
Looks good. Feel free to merge.